GVFS/GVFS.Common/Database/GVFSDatabase.cs

```diff
@@ -0,0 +1,135 @@
+            this.sqliteConnectionString = SqliteDatabase.CreateConnectionString(this.databasePath);
+
+            string folderPath = Path.GetDirectoryName(this.databasePath);
+            fileSystem.CreateDirectory(folderPath);
+
+            for (int i = 0; i < InitialPooledConnections; i++)
+            {
+                this.connectionPool.Add(this.OpenNewConnection());
+            }
+
+            this.Initialize();
+            this.CreateTables();
+        }
+
+        public interface IPooledConnection : IDisposable
+        {
+            SqliteConnection Connection { get; }
+        }
+
+        public void Dispose()
+        {
+            this.disposed = true;
+            this.connectionPool.CompleteAdding();
+            while (!this.connectionPool.IsCompleted && this.connectionPool.TryTake(out SqliteConnection connection))
+            {
+                connection.Dispose();
+            }
+        }
+
+        public IPooledConnection GetPooledConnection()
+        {
+            SqliteConnection connection;
+            if (!this.connectionPool.TryTake(out connection, millisecondsTimeout: MillisecondsWaitingToGetConnection))
+            {
+                connection = this.OpenNewConnection();
+            }
+
+            return new GVFSConnection(this, connection);
+        }
+
+        private void Initialize()
+        {
+            using (IPooledConnection pooled = this.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                command.CommandText = $"PRAGMA journal_mode=WAL;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA cache_size=-40000;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA synchronous=NORMAL;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA user_version;";
+                object userVersion = command.ExecuteScalar();
+                if (userVersion == null || Convert.ToInt64(userVersion) < 1)
+                {
+                    command.CommandText = $"PRAGMA user_version=1;";
+                    command.ExecuteNonQuery();
+                }
+            }
+        }
+
+        private void CreateTables()
+        {
+            using (IPooledConnection pooled = this.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Placeholders.CreateTable(command);
+            }
+        }
+
+        private SqliteConnection OpenNewConnection()
+        {
+            SqliteConnection connection = new SqliteConnection(this.sqliteConnectionString);
+            connection.Open();
+            return connection;
+        }
+
+        private void ReturnToPool(SqliteConnection connection)
+        {
+            if (this.disposed)
+            {
+                connection?.Dispose();

```

Comment:
Does you use of `?.` imply that `connection` could be `null`?

If the connection could be `null` should we check if it's `null` before calling `this.connectionPool.TryAdd`?

Reply:
In looking a the code closer I doesn't appear that this could ever be `null` so I am removing the `?`